modules/internal/ChapelTuple.chpl

```diff
@@ -209,29 +209,35 @@ module ChapelTuple {
     const ignoreRunning = dataParIgnoreRunningTasks;
     const minIndicesPerTask = dataParMinGranularity;
     const length = this.size;
-    const myRange = 1..#length;
+    const myRange = (0..#length,);
     var (numChunks, _) = _computeChunkStuff(numTasks, ignoreRunning,
                                             minIndicesPerTask,
-                                            (myRange,));
+                                            myRange);
 
     if numChunks == 1 {
       yield myRange;
     } else {
-
       coforall chunk in 0..#numChunks {
         // _computeBlock assumes 0-based ranges
         const (lo,hi) = _computeBlock(length, numChunks, chunk, length-1);
-        // adjust for 1-based indexing
-        yield lo+1..hi+1;
+        yield (lo..hi,);
       }
     }
   }
 
   pragma "no doc"
-  iter _tuple.these(param tag:iterKind, followThis)
+  iter _tuple.these(param tag:iterKind, followThis: _tuple)
       where tag == iterKind.follower
   {
-    for i in followThis do yield this(i);
+    // Received followThis from a multi-dimensional object (e.g. a 2D array)
+    if followThis.size != 1 then
+      compilerError('Tuple cannot be zipped with a multi-dimensional object in a forall');
```

Comment:
Will add error message tests. Let's try to narrow down a good error message - I'll ask for opinions from a broader audience.

Here are some ideas:

- "Tuple iteration is incompatible with leading iterator"
- "Iterators are Incompatible"
- "Tuple zipped with incompatible iterator" 
  - since an error would occur regardless of which iterator is leader

Reply:
I think the message that is here isn't so bad, and is attractive in that it's more specific and therefore easier to understand than the other proposals here.  To address @benharsh's concern, the iterator itself could potentially be tightened up to constrain it to only match against yielded tuples of ranges (I'm assuming that Ben's concern is that an associative domain of 2-tuples could/would have a leader that yielded 2-tuples and therefore would match against this call?  Note that the .translate() line is also likely to blow up in weird ways in that case...  I also suspect that many other follower iterators would have similar problems in cases like that...).

If we were to stick with the current assumption ("we're generating an error for cases that try to zip with multidimensional arrays"), I'd reword slightly to "tuples cannot be zipped with multidimensional objects."  If we were to go with something more general, I'd probably say "tuple zipped with incompatible iterator expression."

Reply:
My concern is mainly that ``followThis.size`` or ``followThis(1).type`` doesn't really say anything about what the leader is, at least in practice today. Until we have a better way of determining what the leader is, I'd prefer we have a more generic error message like your last suggestion.

I think it's up for debate whether a generic error message would be helpful for a user, but it could certainly be less confusing than mentioning multidimensional arrays where none may be involved.

Reply:
> but it could certainly be less confusing than mentioning multidimensional arrays where none may be involved.

I agree. I simply hadn't considered irregular array types as other triggers to this compiler error. I would be fine with the error message: "tuple zipped with incompatible iterator expression."